Explain repo and declare all model implementation (#23)
JuliaDynamics/ABM_Framework_Comparisons: README.md

# Agent based modelling frameworks comparison

> This repository contains code used to compare performance and features between various agent based modelling **(ABM)** frameworks. Currently, frameworks compared are [Agents.jl](https://github.com/JuliaDynamics/Agents.jl), [NetLogo](https://ccl.northwestern.edu/netlogo/), [MASON](https://journals.sagepub.com/doi/10.1177/0037549705058073) and [Mesa](https://github.com/projectmesa/mesa). We happily welcome more frameworks to join the comparison.

Reviewer: here I would change the links of Mason and NetLogo for consistency

```suggestion
This repository contains code used to compare performance and features between various agent based modelling **(ABM)** frameworks. Currently, frameworks compared are [Agents.jl](https://github.com/JuliaDynamics/Agents.jl), [NetLogo](https://github.com/NetLogo/NetLogo), [MASON](https://github.com/eclab/mason) and [Mesa](https://github.com/projectmesa/mesa). We happily welcome more frameworks to join the comparison.
```

> **This repository establishes objectively that Agents.jl is the fastest open source agent based modelling framework and that Agents.jl has the simplest source code for a given model specification.**

Reviewer: maybe this is better

```suggestion
**This repository establishes objectively that Agents.jl is the fastest open source agent based modelling framework and that Agents.jl has the simplest source code for a given model specification in comparison to all the other tested frameworks.**
```

**The performance benchmark comparison is run automatically during continuous integration, and hence the comparison is updated after every pull request to this repo.**

## How it works

Various agent based models have been selected to compare performance, such as the Schelling model for example. This repository is structured as follows

1. Each selected ABM is contained in a dedicated folder of this repo.
1. Inside the ABM folder there is a DECLARATION.md markdown file. In declares both the scientific as well as technical implementation of the ABM. We tried our best to make the declaration as clear and as specific as possible, but we welcome Pull Requests that may clarify the declaration even more.
1. In the same folder there are subfolders named after the frameworks. Each contains the files that implement and benchmark the ABM implementation.
1. The implementation must be written in the same way a typical user will use the respective software. The implementations must only use the documented API of the respective software.

> 1. The benchmark step operates as follows: all models must be seeded with a given random number generator seed. At the start of the process, _`S` random seeds are generated in a reproducible way (or, alternatively, a random number generator that generates seeds is initialized with a specified seed)_. `S` is the amount of random seeds and hence also the amount of simulations performed for a given model. Unless stated otherwise in the declaration file of an ABM, `S` has been arbitrarily decided to be `1000`.
> 1. From these `1000` random (but reproducibly random) model runs, the median is used as the performance of each software.

Reviewer: I'm also thinking that 1000 repetitions of each benchmark are too many, I would stick with 100 runs for two reasons:

- I verified that medians are pretty stable even if we use only 100 runs
- We will probably hit the limit of 6 hours for a job running so many repetitions

```suggestion
1. The benchmark step operates as follows: all models must be seeded with a given random number generator seed. At the start of the process, _`S` random seeds are generated in a reproducible way (or, alternatively, a random number generator that generates seeds is initialized with a specified seed)_. `S` is the amount of random seeds and hence also the amount of simulations performed for a given model. Unless stated otherwise in the declaration file of an ABM, `S` has been arbitrarily decided to be `100`.
1. From these `100` random (but reproducibly random) model runs, the median is used as the performance of each software.
```

Reviewer: So it wouldn't be so arbitrare after all :D

# move to repo folder
cd ABM_Framework_Comparisons
```

> If you are using WSL make sure that you move to a folder inside the subsystem before running this commands.

Author: ```suggestion
If you are using WSL make sure that you move to a folder inside the subsystem before running these commands.
```
